Main.cc

```diff
@@ -46,60 +45,81 @@ static bool verbose;
 static bool perflog;
 static int grpc_port(-1);
 
-// Called on connection. Creates session objects and assigns UUID to it
-void OnConnect(uWS::WebSocket<uWS::SERVER>* ws, uWS::HttpRequest http_request) {
+// Sessions map
+std::unordered_map<uint32_t, Session*> sessions;
+
+// Apply ws->getUserData and return one of these
+struct PerSocketData {
+    uint32_t session_id;
+    string address;
+};
+
+void OnUpgrade(uWS::HttpResponse<true>* http_response, uWS::HttpRequest* http_request, struct us_socket_context_t* context) {
     session_number++;
     // protect against overflow
     session_number = max(session_number, 1u);
 
-    uS::Async* outgoing = new uS::Async(websocket_hub.getLoop());
-
-    outgoing->start([](uS::Async* async) -> void {
-        Session* current_session = ((Session*)async->getData());
-        current_session->SendPendingMessages();
-    });
-
     string address;
-    auto ip_header = http_request.getHeader("x-forwarded-for");
-    if (ip_header) {
-        address = ip_header.toString();
+    auto ip_header = http_request->getHeader("x-forwarded-for");
+    if (!ip_header.empty()) {
+        address = ip_header;
     } else {
-        address = ws->getAddress().address;
+        address = http_response->getRemoteAddress();
     }
 
     // Check if there's a token
     if (!auth_token.empty()) {
-        auto token_header_entry = http_request.getHeader("carta-auth-token");
-        if (token_header_entry) {
-            string token_header_value = token_header_entry.toString();
+        auto token_header_entry = http_request->getHeader("carta-auth-token");
+        if (!token_header_entry.empty()) {
+            string token_header_value(token_header_entry);
             if (token_header_value != auth_token) {
-                ws->close(1000, "Header auth failed");
+                std::cerr << "Header auth failed!\n";
```

Comment:
I think we've standardised on `fmt::print` for these messages.

Reply:
There are many std::cerr in the code, and many fmt::print.  I was not aware of this standard.

Reply:
Hmm. I could have sworn that we made an attempt to switch over at some point in the past -- it's why we added `fmt` as a dependency. It does look very inconsistent right now. I also noticed that we're using `fmt::print` for error messages without specifying `stderr` as the stream, which prints them to `stdout`.

I think we should make a separate issue for this, and maybe discuss it in the next Slack meeting.

Reply:
I agree.